driver-core/src/main/com/mongodb/client/model/expressions/ArrayExpression.java

```diff
@@ -50,25 +50,30 @@
      */
     <R extends Expression> ArrayExpression<R> map(Function<? super T, ? extends R> in);
 
+    ArrayExpression<T> sort();
+
     IntegerExpression size();
 
-    BooleanExpression any(Function<T, BooleanExpression> map);
+    BooleanExpression any(Function<T, BooleanExpression> mapper);
+
+    BooleanExpression all(Function<T, BooleanExpression> mapper);
 
-    BooleanExpression all(Function<T, BooleanExpression> map);
+    NumberExpression sum(Function<T, NumberExpression> mapper);
 
-    NumberExpression sum(Function<T, NumberExpression> map);
+    NumberExpression multiply(Function<T, NumberExpression> mapper);
 
-    NumberExpression multiply(Function<T, NumberExpression> map);
+    <R extends Expression> R max(Function<T, R> mapper, R orElse);
 
-    NumberExpression max(Function<T, NumberExpression> map, NumberExpression orElse);
+    <R extends Expression> R min(Function<T, R> mapper, R orElse);
```

Comment:
`ArrayExpression.first`/`last` don't take a fallback value, but `min`/`max` do. Since there is no conceptual difference between these two sets of methods, we should use the same approach in them.

Reply:
There is also `elementAt`, `getBoolean(String)` (and all the rest), and `get` on MapExpression. I think my plan is to make the API type-safe wherever I can get away with it, and I think having a default is passable here but more problematic elsewhere.

Reply:
Currently, the `ArrayExpression.min` implementation is

```java
this.size().eq(of(0)).cond(
    other,
    results.minN(of(1), v -> v).first())
```

What stops us from doing a similar thing in `first`?

```java
private T unsafeFist() {
    return new MqlExpression<>(astWrapped("$first")).assertImplementsAllExpressions();
}

@Override
public T first(final T other) {
    return this.size().eq(of(0)).cond(
            other,
            unsafeFist());
}
```

and then, of course, `ArrayExpression.min` becomes

```java
@Override
public <R extends Expression> R min(final R other, final Function<? super T, ? extends R> mapper) {
    MqlExpression<R> results = (MqlExpression<R>) this.map(mapper);
    return this.size().eq(of(0)).cond(other, ((MqlExpression<R>)results.minN(of(1), v -> v)).unsafeFist());
}
```

I tried it and it even works. It expectedly changes the MQL behind `ArrayExpression.first`, but the MQL for `ArrayExpression.min` stays unchanged.

Reply:
I think we discussed this ("securing" the API), but agreed that it was impractical to have everywhere due to performance, and therefore not worthwhile here (due to the additional inconvenience of needing to specify the other value). [Comment here](https://github.com/mongodb/mongo-java-driver/pull/1032/files#r1048611098).

Reply:
If the conditional logic is deemed to introduce too much overhead in `ArrayExpression.first`/`last`, then why is it not deemed to have the same affect in `ArrayExpression.min`/`max`? I see no relevant difference between these sets of methods, which is why I think that the same approach should be used in both.

Reply:
First and last are derivations of elementAt, which is the array version of `get` for maps/records. These are all conceptually a single method. It seems unavoidable that users would need to assert for variations of `get` (getInteger, getArray...), since (as we have discussed) forcing a default value there seems like it would greatly slow down array traversal, which is a very common usage.

But this consideration does not seem to exist for min/max, because the check is applied only once for an array rather than once per element. So we are not compelled to offer a second method for this category of methods, especially a second method that is unsafe.

Reply:
Discussed this over Zoom, and while doing so, the following realization came to mind: as long as we tend to value performance over safety (which so far was the case for this API), `min`/`max` methods are different from `first`/`last` in that they are much more computationally intensive (when called on an unsorted array, they have linear time complexity, and if the array is sorted, then hopefully the server optimizes them, but I don't know if that's what happens). Adding a trivial conditional logic to a relatively heavy `min`/`max` is fine, as it does not significantly affect their performance. But adding the same check to an extremely lightweight constant-time methods `first`/`last` may roughy double the amount of work these methods do.